tools/fragpipe/fragpipe.xml

```diff
@@ -0,0 +1,286 @@
+    <expand macro="requirements"/>
+
+    <stdio>
+        <!-- The error code from individual tools run by FragPipe is not always propogated to the FragPipe process exit code -->
+        <regex match="exit code: [^0]"
+               source="stdout"
+               level="fatal"
+               description="A FragPipe pipeline process returned non-zero exit code."/>
+    </stdio>
+
+    <version_command>fragpipe --help | grep 'FragPipe v'</version_command>
+
+    <!--
+        FragPipe wrapper command:
+        - Links input files
+        - Modifies database path in workflow file
+        - Runs Philosopher to add decoys and contaminants
+        - Runs FragPipe, passing paths to MSFragger, Philosopher, Python, and IonQuant
+    -->
+    <command detect_errors="exit_code"><![CDATA[
+        mkdir .cache .config tmp &&
+        export XDG_CONFIG_HOME=\$(pwd)/.config &&
+        export XDG_CACHE_HOME=\$(pwd)/.cache &&
+        export JAVA_OPTS="\$JAVA_OPTS -Xmx\$[ \${GALAXY_MEMORY_MB:-8192} / 1024 ]g -Djava.io.tmpdir=\$(pwd)/tmp" &&
+        export LD_LIBRARY_PATH="\$LD_LIBRARY_PATH:/usr/local/lib" &&
+        mkdir -p outputs &&
+        @CMD_IMPORTS@
+        @LINK_FRAGGER_DATABASE@
+        @LINK_SAMPLES@
+        @UPDATE_WORKFLOW@
+        #if not $workflow_only
+            #if str($database_options) != 'none'
+                philosopher workspace --init &&
+                '$__tool_directory__/genericize_db.py' ${database_name} genericized_db.fas &&
+                philosopher database --custom 'genericized_db.fas' --prefix rev_
+                #if str($database_options) != 'decoy_and_contam'
+                    --contam
+                #end if
+                --contamprefix &&
+                mv *-decoys-*.fas decoys-contam-database.fas &&
+            #end if
+            fragpipe --msfragger_key f28c4be2-759b0a11-6c1ea652-1240565a --ionquant_key 47cb2953-b4bf7726-2c96f5fe-8f1013aa --headless --threads $threads --ram \$[ \${GALAXY_MEMORY_MB:-8192} / 1024 ] --workflow fp.workflow --manifest fp.manifest --workdir outputs
+            --config-msfragger \${CONDA_PREFIX}/share/msfragger-*/MSFragger-*/MSFragger-*.jar --config-philosopher `which philosopher` --config-python `which python` --config-ionquant \${CONDA_PREFIX}/share/ionquant-*/IonQuant.jar
+            && cd outputs
+            && ln -s log_*.txt log.txt
+            #if $output_options and 'combined_outputs' in $output_options
+                && mkdir combined_outputs
+                && for f in combined*tsv; do [[ -e $\{f} || continue ]]; mv \${f} combined_outputs; done
+            #end if
+            && for f in psm.tsv ion.tsv peptide.tsv protein.tsv; do n=0; for i in `find * -name \${f}`; do d=\${i%/*}; echo "\${n}  \${d} \${i}"; if [[ \$n == 0 ]]; then awk 'NR==1{printf("Experiment\t%s\n",\$0);}' \$i > concat_\${f}; fi; awk 'NR>1{printf("'"\${d}"'\t%s\n",\$0);}' \$i >> concat_\${f}; t=`echo \${i} | tr '/' '.'`; mv \${i} \${t}; n=\$((n+1)); done; done
```

Comment:
Can you describe what you are doing here, e.g. in a code comment? My preference would also be to indent the bash loop and if statements to make this readable.

Reply:
I updated up the entire `<command>` tag to be more clear in this commit (https://github.com/galaxyproteomics/tools-galaxyp/pull/761/commits/5c2189e00f8aa1a451b567d51c16c4a252077f62), including indenting and commenting this. I broke `<command>` into further macros as well.

Copying the comment I added:
> For each category of results - psm, ion, peptide, protein - this loop concatenates results from each sample into a single result. For example, the entries sample1/psm.tsv and sample2/psm.tsv would be combined to concat_psm.tsv, with a single psm.tsv header as its first line.